src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
+                    return dispatch::make_ready_future();
+                });
+        });
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::UploadFile(const std::string& project_id,
-                                                 const std::wstring& file,
-                                                 const Language& lang,
+dispatch::future<void> CrowdinClient::UploadFile(const long project_id,
+                                                 const long file_id,
+                                                 const std::string& lang_tag,
                                                  const std::string& file_content)
 {
-    auto url = "/api/project/" + project_id + "/upload-translation";
-
-    multipart_form_data data;
-    data.add_value("json", "");
-    data.add_value("language", lang.LanguageTag());
-    data.add_value("import_duplicates", "0");
-    data.add_value("import_eq_suggestions", "0");
-    data.add_file("files[" + str::to_utf8(file) + "]", "upload.po", file_content);
-
-    return m_api->post(url, data).then([](json){});
+    return RefreshToken().then([=] () {
+    return m_api->post(
+            "storages",
+            octet_stream_data(file_content),
+            { { "Crowdin-API-FileName", L"poedit.xliff"} }
+        )
+        .then([this, project_id, file_id, lang_tag] (json r) {
+            cout << "File uploaded to temporary storage: " << r << "\n\n";
+            const auto storageId = r["data"]["id"].get<int>();
+            return m_api->post(
+                "projects/" + std::to_string(project_id) + "/translations/" + lang_tag,
+                json_data(json({
+                    { "storageId", storageId },
+                    { "fileId", file_id },
+                    { "importDuplicates", true },
+                    { "autoApproveImported", true }
+                })))
+                .then([](json r) {
+                    cout << "File uploaded: " << r << "\n\n";
+                });
+        });
+    });
 }
 
-
-bool CrowdinClient::IsSignedIn() const
+json CrowdinClient::LoadAuth()
 {
-    std::string token;
-    return keytar::GetPassword("Crowdin", "", &token);
+    std::string auth;
+    return keytar::GetPassword("Crowdin", "", &auth)
+            // if not came from older (<2.4) version
+            && auth[0] == '{'
+                ? json::parse(auth)
+                : json();
 }
 
-
-void CrowdinClient::SignInIfAuthorized()
+void CrowdinClient::SetAuth(const json& auth)
 {
-    std::string token;
-    if (keytar::GetPassword("Crowdin", "", &token))
-        SetToken(token);
-}
+    cout << endl << endl << "Authorization: " << auth << endl << endl;
+
+    if(auth.empty()) {
+        return;
+    }
 
+    std::string access_token = auth["access_token"];
+    m_authRefreshToken = auth["refresh_token"].get<string>();
+    m_authExpireTime = auth["expiration_time"].get<time_t>();
+    
+    auto domain = jwt::decode(access_token).get_payload_claim("domain");
```

Comment:
This seems to be the only place where the jwt library is used.

I'd rather not add two dependencies just for this. One of them in particular, OpenSSL, is something the commercial branch of Poedit already used and I'm trying hard to get rid of it (in favor of libsodium) - it's a big dependency for something this small. (And the other one adds *another* C++ JSON parser.)

I'd like to propose two possible ways forward:

(1) Ideally, the simplest would be for Crowdin to add the domain as data alongside the returned tokens as another JSON key. It's my understanding that the v2 API is still in beta, so perhaps they could be persuaded?

(2) Failing that, if I look at JWT spec, the token is simply 3 chunks of base64-encoded data. Since you don't verify the signature (a good call when talking over HTTPS), no crypto is needed, and you can get the domain payload by base64-decoding the middle chunk and parsing it as JSON.

Either way that's easily 1MB of binary size saved for equivalent functionality compared to pulling in the two dependencies.

Reply:
> This seems to be the only place where the jwt library is used.

Yep, correct

> I'd rather not add two dependencies just for this. One of them in particular, OpenSSL, is something the commercial branch of Poedit already used and I'm trying hard to get rid of it (in favor of libsodium) - it's a big dependency for something this small.

Just FYI (if you did not look too deep to notice but if it would affect your opinion exactly in this case) - there are header only dependency for both `jwt-cpp` and `openssl`, i.e. no library building/linking required for neither of them in this case. And they already added to the build workflow on all platform in this PR (i.e. as submodules to `deps`, their headers etc.)

> And the other one adds another C++ JSON parser.

I dislike this too thought it's extremely lightweight

And anyway I also prefer less dependencies so understand your concern so will do my best to eliminate them here. 

> (1) Ideally, the simplest would be for Crowdin to add the domain as data alongside the returned tokens as another JSON key. It's my understanding that the v2 API is still in beta, so perhaps they could be persuaded?

I asked them for exactly this already but they asking "why to modify API adding something more to response which just duplicates that already available in JWT).

> (2) Failing that, if I look at JWT spec, the token is simply 3 chunks of base64-encoded data. Since you don't verify the signature (a good call when talking over HTTPS), no crypto is needed, and you can get the domain payload by base64-decoding the middle chunk and parsing it as JSON.

It definitely makes sense so I will look into this to implement as you propose if you think that current headers-only usage of above libs are certainly unacceptable in our case.  Or may be will find some another C++ JWT lib which is much simple just enough for this purpose without any dependencies.

> Either way that's easily 1MB of binary size saved for equivalent functionality compared to pulling in the two dependencies.

Did you notice so big size increase of new version? Since compiler usually optimizes by taking from headers-only libs only what is really used what it directly depends on.

Reply:
> Did you notice so big size increase of new version?

No, sorry, I simply didn't realize OpenSSL was *actually* not used by the execution path and not linked in at all. That 1MB figure comes from other uses in Poedit Pro where it was ~minimal achievable size do to its own spaghetti internal dependencies.

Reply:
Addressed in #633 b4a90b7db57d06b5174215f69e9f1be5c35d9ed6 exactly as discussed